cloudfoundry/resource_cf_app.go

```diff
@@ -999,7 +1211,22 @@ func prepareApp(app cfapi.CCApp, d *schema.ResourceData, log *cfapi.Logger) (pat
 	return path, nil
 }
 
-func validateRoute(routeConfig map[string]interface{}, route string, rm *cfapi.RouteManager) (routeID string, err error) {
+func validateRoute(appID string, routeID string, rm *cfapi.RouteManager) error {
+	if mappings, err := rm.ReadRouteMappingsByRoute(routeID); err == nil && len(mappings) > 0 {
+		if len(mappings) == 1 {
+			if boundApp, ok := mappings[0]["app"]; ok && boundApp == appID {
+				return nil
+			}
+		}
+		return fmt.Errorf(
+			"route with id %s is already mapped. routes specificed in the 'routes' argument can only be mapped to one 'cf_app' resource",
```

Comment:
I carried this restriction over from the original routes code.  I feel that it still makes sense because if you want to manage a route in this fashion, you should use the `cf_route` resource for this purpose. 
Furthermore, once we get into implementing blue/green support, the complexities of a proper blue/green deployment process in CF require that the `cf_app` resource is aware of all routes mapped to the application that need to continue to always route to a live application during the update, and those routes probably need to be exclusive to that application in order to make "blue/green sense".

Reply:
@jcarrothers-sap 

It seems that this check is currently only performed on mapped routes during updates, as the following TF syntax at https://gist.github.com/gberche-orange/59d9075896e69313a145a44ae04e592c  indeed successfully resulted in mapping a given route to multiple apps with 0.9.9 version : 

``` 
cf a
Getting apps in org bercheg-org / space demo-basel as xx..
OK

name    requested state   instances   memory   disk   urls
dora1   started           1/1         1G       1G     dora-tf-demo.cfapps.io
dora2   started           1/1         1G       1G     dora-tf-demo.cfapps.io
```

The provider only failed after trying to further update app routes, as the restriction seems only applied when reading the existing route mapping. See related traces at https://gist.github.com/gberche-orange/c651aa2743faa6c958cae10b3ffe2524


> I carried this restriction over from the original routes code
>  I feel that it still makes sense because if you want to manage a route in this fashion, you should use the cf_route resource for this purpose.
> Furthermore, once we get into implementing blue/green support [...]

We'll have to see whether we indeed need blue/green support in the terraform provider, or whether a "zero downtime" is sufficient, see #25 and https://github.com/cloudfoundry/cloud_controller_ng/blob/master/docs/v3/zero_downtime_deployments.md

I still have preferences for:
1- removing this consistency check along with old route management
2- if we transiently keep it, document it in the `cloudfoundry_application` documentation

We can however live with this limitation if we don't find time to prioritize this, as we need to get 1.0 out asap. Feel free then to resolve this conversation if you feel we should tackle this post 1.0

Reply:
@gberche-orange 

>It seems that this check is currently only performed on mapped routes during updates, as the following TF syntax at https://gist.github.com/gberche-orange/59d9075896e69313a145a44ae04e592c  indeed successfully resulted in mapping a given route to multiple apps with 0.9.9 version : 

Based on an examination of the code, I believe that the reason your sample TF syntax works is due to a race condition as both applications would likely end up being created in parallel, but I am certain that the current create code does check and attempt to enforce the exclusivity of routes when they are mapped using the `cloudfoundry_app` resource as I have experienced this limitation in the past working with the older code base.

> We'll have to see whether we indeed need blue/green support in the terraform provider, or whether a "zero downtime" is sufficient, see #25 and https://github.com/cloudfoundry/cloud_controller_ng/blob/master/docs/v3/zero_downtime_deployments.md

It's possible the new "zero downtime" feature may be sufficient, but I do not believe it is yet available for my team's deployments of CF, and we are currently using an internal fork of the provider that has a preliminary, but working, implementation of the blue/green feature implemented in the provider.

> I still have preferences for:
> 1- removing this consistency check along with old route management

After thinking about this some more, I have realised that the exclusivity has more to do with the convention that Terraform takes full ownership of resources and is normally responsible for correcting configuration drift.  Without the need for blue/green this wouldn't be an issue as defining route mappings on the `cloudfoundry_app` resource wouldn't be implemented and we could just rely on using `cloudfoundry_route` resources to control the mappings. Then there wouldn't be any question about correcting configuration drift.  However, due to the thought that the provider was going to have to implement blue/green support on its own, this led to the need to be able to manage route mappings from the app resource.

> 2- if we transiently keep it, document it in the `cloudfoundry_application` documentation

I agree that this limitation should be documented.  

> We can however live with this limitation if we don't find time to prioritize this, as we need to get 1.0 out asap. Feel free then to resolve this conversation if you feel we should tackle this post 1.0

I propose that we discuss this in detail at the sync-up meeting on Monday.